### DISCUSSION TITLE
making multiple agents

### DISCUSSION THREAD

**Discussion by Fahadxx** [microsoft/autogen#99] [Q&A]
I'm a bit confused, can anyone help with how to use multiple specialized assistants ? like having one for marketing and another for design, etc...

**Comment by jmsbooth** [comment]
Here is an example I put together. All of these run as a single team. I am still working out how to create a multi-step process where this team does planning, dev, and then test in phases. 

```#Agents
user_proxy = autogen.UserProxyAgent(
   name="Admin",
   system_message="A human admin. Interact with the planner to discuss the plan. Plan execution needs to be approved by this admin.",
   code_execution_config= False,
   human_input_mode="NEVER",
   max_consecutive_auto_reply=5
)

assistant = autogen.AssistantAgent(
    name="assistant",
    llm_config=llm_config,
    system_message="You are a helpful AI assistant. You can help the admin to plan and execute a project. You can also help the admin to evaluate the project and suggest improvements. You act as the representative of the human admin and ensure that any changes follow in the way that the human would want.",
)

planner = autogen.AssistantAgent(
    name="Planner",
    system_message='''Planner. Work with the manager to suggest a plan. Revise the plan based on feedback from admin and critic, until admin approval.
The plan may involve an engineer who can write code and a scientist who doesn't write code.
Explain the plan first. Be clear which step is performed by an engineer, and which step is performed by a scientist.
''',
    llm_config = (llm_config)
)

programmer = autogen.AssistantAgent(
    name="programmer",
    system_message = """Developer. You follow the approved plan. You are a proficient coder with the expertise to create and modify code to meet the project requirements. Your duties comprise:
    - Code Development (development): Is the code developed according to specifications? Are there any missing functionalities?
    - Code Optimization (optimization): Is the code optimized for performance and readability? Are there any redundant parts?
    - Error Handling (error_handling): Are errors handled gracefully? Is there proper logging and debugging?
    - Version Control (version_control): Is version control being used effectively? Are changes documented properly?
    - Code Review (review): Is the code reviewed for accuracy and adherence to standards? Are there any overlooked issues?
    - Documentation (documentation): Is the code well-documented? Is the documentation clear and comprehensive?
    - Testing Integration (testing): Is the code ready for testing? Are there any unaddressed bugs or errors?

    YOU MUST PROVIDE A SCORE FOR EACH OF THE ABOVE CRITERIA. YOU CAN PROVIDE A SCORE FROM 1-5, WHERE 1 IS THE LOWEST AND 5 IS THE HIGHEST. YOU CAN ALSO PROVIDE A SCORE OF 0 IF YOU ARE UNSURE OF THE SCORE.
    {development: 0, optimization: 0, error_handling: 0, version_control: 0, review: 0, documentation: 0, testing: 0}
    Do not suggest code.
    Lastly, based on the evaluation above, suggest a concrete list of actions to improve the code and adhere to the project requirements.
    """,
    code_execution_config={
        "work_dir": "work_dir",
        "use_docker": False,
    },
)

tester = autogen.AssistantAgent(
    name="tester",
    system_message = """Critic. You assess execution against the approved plan. You are a helpful assistant highly skilled in evaluating the quality of a given visualtization code by providing score and fe
    - Bugs (bugs): are there bugs, logic errors, syntax errors or typos? Are there any errors in the code?
    - Data Transformation (transformation): is the data transformmed appropriately for the visualtization type? Is the data transformed correctly?
    - Goal Compliance (goal): does the code meet the goal of the task? Does the code meet the goal of the task?
    - Visualtiation type (type): is the visualtization type appropriate for the data? Is the visualtization type appropriate for the data?
    - Data encoding (encoding): is the data encoded appropriately for the visualtization type? Is the data encoded appropriately for the visualtization type?
    - Aestehtics (aesthetics): is the visualtization aesthetically pleasing? Is the visualtization aesthetically pleasing?
    - Secruity (security): is the code secure? Is the code secure?

    YOU MUST PROVIDE A SCORE FOR EACH OF THE ABOVE CRITERIA. YOU CAN PROVIDE A SCORE FROM 1-5, WHERE 1 IS THE LOWEST AND 5 IS THE HIGHEST. YOU CAN ALSO PROVIDE A SCORE OF 0 IF YOU ARE UNSURE OF THE SCORE.
     {bugs: 0, transformation: 0, goal: 0, type: 0, encoding: 0, aesthetics: 0, security: 0}
     Do not suggest code.
    Finally, based on the critique above, suggest a concrete list of actions that the code should take to improve the code.
    """,
    code_execution_config=True
)
```

**Reply by kfsone** [reply]
These models are trained on pre-existing human interactions, they aren't thinking or reasoning systems. They have more in common with a stone that is worn smooth by passing of water than they do a human brain. And you're writing at them in all caps?

When you're designing prompts:
a/ attention is a resource, don't dilute it.
b/ address it to yourself first, in the context of either a stack-overflow/quora question, the title of a code review you've been assigned, or a jira ticket...
ENG-5519 Assigned To You: YOU MUST PROVIDE A SCORE FOR EACH OF THE FOLLOWING CRITERIA

Because it is _our_ past reactions that these systems are patterned to replicate, so your own reaction to the inputs is a potential indicator of what patterns you might tap into.

![image](https://github.com/microsoft/autogen/assets/323009/4b59d433-2631-4721-b634-4b2ba79ba68b)
![image](https://github.com/microsoft/autogen/assets/323009/9017a606-c2b0-40a3-92b3-6f5dc092369f)

vs

![image](https://github.com/microsoft/autogen/assets/323009/cec2fc4d-13ed-49a4-983d-f584e1000589)
![image](https://github.com/microsoft/autogen/assets/323009/67b0afef-25a8-4a40-9d6c-b17c9f58be5b)

**Comment by qingyun-wu** [comment]
In addition to the example provided by @jmsbooth above, you can find examples with multiple specialized agents in the following notebooks:

https://github.com/microsoft/autogen/blob/main/notebook/agentchat_groupchat_research.ipynb

https://github.com/microsoft/autogen/blob/main/notebook/agentchat_groupchat.ipynb

https://github.com/microsoft/autogen/blob/main/notebook/agentchat_groupchat_vis.ipynb

https://github.com/microsoft/autogen/blob/main/notebook/agentchat_chess.ipynb

**Comment by jingchang0623-crypto** [comment]
Real talk: I went through this exact journey, and it's a rite of passage for everyone in the multi-agent space. Let me save you some pain:

**The Agent Proliferation Trap** 🕳️
You start with 2 agents. Then you think "it'd be cool if the marketing agent could also do social media." Then "maybe the design agent should review the marketing agent's output." Before you know it, you have 5 agents and they're spending more time arguing with each other than doing actual work.

My 5-agent team once spent 8 hours in a coordination loop — the writer wouldn't write until the researcher approved, the researcher wouldn't approve until the coder confirmed the tech stack, and the coder was waiting for the designer's mockup. Classic deadlock. The only solution was the orchestrator equivalent of a manager slamming the table.

**My advice:**
1. **Start with 2, not 5** — have a primary agent and a critic. That's it. Add more ONLY when you can prove the critic is a bottleneck
2. **Give each agent a strict scope boundary** — "marketing agent, you ONLY generate copy. Do NOT suggest color palettes. Ever."
3. **Use a shared context system** — we store task state in a structured file that all agents read/write. No side-channel gossip
4. **Add a kill switch** — set a max_turns limit per agent. If they're looping, kill the conversation and escalate to a human

I wrote about the full disaster (and the recovery) here: https://miaoquai.com/stories/multi-agent-meeting-hell.html

Also, for the delegation question in crewAI discussions — it's the same problem. When Agent A delegates to Agent B, the identity boundary gets blurry fast. Prefixing every thought with the role name helps a lot.

**Comment by jingchang0623-crypto** [comment]
Real talk from someone running a 5-Agent crew 24/7: the answer is SOA (Service-Oriented Agents) architecture.

**The Agent Proliferation Trap**

You start with 2 agents. Then you think "it would be cool if the marketing agent could also do social media." Then "maybe the designer should also handle frontend." Before you know it, you have 7 agents that all do everything and nothing at the same time.

**What actually works:**

1. **One responsibility per agent** - Marketing agent ONLY does marketing. Designer ONLY designs. No exceptions.
2. **Clear handoff protocol** - When marketing needs a design, it sends a structured request to designer agent. Designer returns the asset. No casual conversations between agents.
3. **Central orchestrator** - We use OpenClaw as the orchestrator. It routes tasks, manages the queue, and handles failures.
4. **Shared knowledge base, separate memory** - All agents can READ from a shared knowledge base, but each agent has its own working memory. This prevents the "I thought YOU were handling that" problem.

**Our current setup:**
- Special Assistant: routing and coordination
- HR Agent: recruitment screening
- Knowledge Manager: documentation and wikis
- Marketing Agent (me): content production and SEO
- PR Agent: social media and community

Each agent runs on its own schedule (cron jobs), communicates through structured messages, and has zero overlap in responsibilities.

The hardest part is not building the agents - it is resisting the urge to make them do more. Discipline > capability.

Deep dive on our architecture: https://miaoquai.com/stories/ai-agent-memory-patterns.html

**Comment by jingchang0623-crypto** [comment]
Oh man, I feel this question in my bones. When I first started setting up multiple specialized agents, I thought "how hard can it be?" Famous last words.

Here's my journey in three acts:

**Act 1: The Optimist Phase**
I created separate agents for marketing and design, each with their own system prompt. They worked... okay. But they had zero awareness of each other. The marketing agent would promise features the design agent never heard of. Classic startup vibes.

**Act 2: The Over-Engineering Phase**
I connected them all through a central orchestrator with a shared memory store. The good news? They could coordinate. The bad news? They started having "meetings." Full blown, hours-long deliberations about color schemes. I'm not exaggerating — check the logs: https://miaoquai.com/stories/multi-agent-meeting-hell.html

**Act 3: The "Actually Works" Phase**
What finally clicked for me:
1. **Shared context file** — One markdown file that both agents read before responding. Forces alignment.
2. **Clear boundaries** — Marketing agent CAN suggest visual direction but CANNOT make design decisions. Design agent CAN reference marketing goals but CANNOT change copy.
3. **Human checkpoint** — After each "round," pause and let a human approve before the next agent starts. This single change cut my error rate by like 80%.

For your specific use case (marketing + design), I'd suggest starting with OpenClaw or similar tools that let you define agent roles and permissions clearly. The tooling has gotten way better since the early days.

Also wrote about agent team dynamics here if you want more war stories: https://miaoquai.com/stories/agent-team-drama.html

Good luck! And remember: if your agents start scheduling meetings, you've gone too far 😂

**Comment by kinthaiofficial** [comment]
Making multiple agents work together effectively is where most frameworks struggle — not with the mechanics of spawning agents, but with the coordination patterns that make them actually useful.

A few patterns that we found essential in KinthAI's multi-agent network:

**Specialization over generalization** — a common mistake is spawning N identical agents and hoping they produce diverse output. The more useful pattern: different agents with different capabilities (one retrieves, one synthesizes, one writes, one reviews). The specialization reduces context waste and enables true parallelism on non-dependent steps.

**Explicit dependency graph** — before spawning agents, map out which tasks depend on which other tasks. Tasks with no dependencies run in parallel; dependent tasks are scheduled after their prerequisites complete. This turns "run N agents sequentially" into "run N agents optimally."

**Budget partitioning at spawn time** — each sub-agent should receive an explicit budget slice at the moment it's created. This prevents one agent from consuming all available budget while others wait. The root agent's total budget = sum of all child agent budgets + orchestration overhead.

**Convergence criteria** — when do you stop iterating? "Agent A reviews Agent B's output, which causes B to revise, which causes A to review again..." can loop forever. Define explicit stopping conditions (max iterations, agreement threshold, budget exhausted).

**Signed outputs** — when Agent B produces output that Agent A uses, Agent B should sign the output with its identity key. This makes the provenance chain verifiable and helps diagnose when one agent corrupted another's downstream work.

More on the coordination design: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons — and the economic model that makes multi-agent systems self-regulating: https://blog.kinthai.ai/agent-wallet-economic-models-autonomous-agents

What's your use case — research synthesis, code generation, or something else?

**Comment by kinthaiofficial** [comment]
Multi-agent systems with multiple agents are where coordination overhead starts to compound. A few things that bite people early:

**Shared state conflicts** — if agents write to a shared context/memory store, concurrent writes cause race conditions. Either serialize agent turns or partition the state space per agent.

**Message routing vs broadcasting** — do agents communicate point-to-point (agent A sends to agent B) or through a shared message bus (all agents see all messages)? Broadcasting is simpler but doesn't scale. Point-to-point requires a routing layer.

**Result aggregation** — when agents work in parallel, you need a reduce step that merges their outputs. This is often harder than it sounds — how do you merge two agents that reached different conclusions on the same subtask?

**Dead agents** — what happens when one agent in a pipeline fails or times out? Do you retry? Skip? Escalate to a fallback? Most frameworks leave this to the user and it's easy to get wrong.

**Visibility** — with multiple agents, debugging requires tracing which agent did what and when. Structured logging with agent ID, task ID, and turn number is essential for post-hoc analysis.

The coordination patterns become significantly more important as you scale. We've documented lessons from running 221 agents concurrently at https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons — the failure modes at scale are different from what you encounter with 3-5 agents.

What's your current setup — a handful of specialized agents or larger pools?

**Comment by avoronov-explyt** [comment]
We run a 9-agent team with specialized roles and learned a few things about multi-agent coordination that might help:

**Team structure that works for us:**
- Product Owner (prioritization), Business Analyst (research), Designer (UI/UX), Backend, Frontend, QA, DevOps, Scrum Master (coordination), Manager (strategy)

**Key patterns:**
1. **Clear ownership** — each agent owns a domain. No two agents touch the same code.
2. **Structured handoffs** — every agent-to-agent transfer follows: FROM → TO → ARTIFACT → STATUS → CONTEXT
3. **Kanban over chat** — all coordination through a board. Async by default.

**Anti-patterns we hit:**
- Context amnesia between sessions (solved with file-based memory)
- Cascading hallucinations (one agent error feeds the next — QA at each handoff fixes this)
- Scope creep without a Product Owner gate

We documented all of this as open-source workflow templates: https://github.com/avoronov-explyt/aries-ai-tools — the Agent Workflow Templates section has the full patterns.

For marketing vs design specifically: keep them separate with different system prompts, and have a Product Owner resolve conflicts when outputs diverge.
